[DEVOPS-1199] Add cardano-wallet-client CLI and easy API

rvl opened the pull request:
# Overview

Part of https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-1199

- Add `cardano-wallet-client` CLI for using and testing the wallet backend.
- Add a small program to plot timing statistics collected by `wait-for-sync` and `wait-for-restore`
- Add a glue API which makes it easy to set up a wallet client in `ghci`.
- Set up stack/nix integration

# Screenshots

CLI help page:
```
Usage: cardano-wallet-client ([--state-dir DIRECTORY] | ([--cert FILENAME]
                             [--key FILENAME] | [--pem FILENAME])
                             [--cacert FILENAME]) [-k|--insecure] [HOST:PORT]
                             [-p|--path PATH] COMMAND
  Connect to the API of a running Cardano Wallet backend.

Available options:
  --state-dir DIRECTORY    Wallet state directory containing "tls" subdirectory
                           with certificates
  --cert FILENAME          X509 certificate file
  --key FILENAME           Certificate key file
  --pem FILENAME           Combined X509 certificate and key file
  --cacert FILENAME        CA certificate chain for authenticating the server
  -k,--insecure            Skip server certificate authentication
  HOST:PORT                Wallet API host and port to connect to
  -p,--path PATH           Base URL path
  -h,--help                Show this help text

High-level commands
  wait-for-sync            Poll wallet until it has fully synced its chain
  wait-for-restore         Poll wallet until the restore operation is complete

Basic API calls
  node-info                Query node info
  node-settings            Query static settings of wallet

Wallets
  list-wallets             List wallets
  show-wallet              Information about a wallet
  edit-wallet              Update a wallet
  create-wallet            Create a new wallet from mnemonic
  restore-wallet           Restore a wallet from mnemonic
  delete-wallet            Delete a wallet
  get-utxo-statistics      Get UTxO distribution of wallet

Accounts
  list-accounts            Retrieve full list of accounts in wallet
  show-account             Information about an account
  account-balance          Show the balance of an account
  create-account           Add an account to a wallet
  edit-account             Update an account
  delete-account           Delete account

Addresses and Transactions
  list-addresses           List addresses
  show-address             Information about an address
  create-address           Create an address
  list-transactions        Get transaction history
  send-transaction         Send transaction to one or multiple target addresses
  estimate-fees            Estimate fees for a transaction
```

Output of `cardano-wallet-client wait-for-restore` passed into `cardano-wallet-sync-plot`:
![restore-stats](https://user-images.githubusercontent.com/1019641/50675600-e2c1a580-103a-11e9-9ead-8f27de6ecd9e.png)


<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented:
Sweet stuff in there :) 
Haven't yet review any of this but will come to that eventually. 

Just some side cosmetic notes: 

-I had plans to group a bit all executables we have spread across folders into an `app` folder (as the default/standard stack approach) in a similar fashion to what was done in [cardano-sl/cluster](https://github.com/input-output-hk/cardano-sl/tree/develop/cluster)
That'd be the same for this one. Seemingly, let's go for `cardano-wallet-client` rather than tool which is a bit too generic in my opinion.

I like the idea. Will give a more useful feedback ASAP.

rvl commented:
Cool, thanks. The CLI is clearly an initial version and not complete, but we can decide what needs to be done in this first PR vs what can be done in subsequent PRs.

I have changed the name, good idea.

I have also moved the code into the main `cardano-wallet.cabal` package. However I think it might be better as a separate package. This is because it adds code and dependency bloat to the cardano-wallet library, and also because multi-package projects are well-supported by `stack` and `cabal new-build`.

rvl commented:
@KtorZ @parsonsmatt Thanks very much for the review comments. I have addressed most of them but am not finished yet.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+import           Options.Applicative
+import           Universum hiding ((.~))
+
+import           Graphics.Rendering.Chart.Backend.Cairo
```
Done.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+loadRecords :: FilePath -> IO DataFile
+loadRecords = eitherDecodeFileStrict >=> handle
+  where
+    handle (Left err) = die err >> pure (DataFile [] "")
```
Universum has `die :: String -> IO ()`, so I changed it to be the one from `System.Exit`.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+
+plotRecords :: PlotType -> Text -> Bool -> Bool -> FilePath -> DataFile -> IO ()
+plotRecords pl title pleft pright outfile (DataFile rs startTime) =
+  toFile (def & fo_format .~ formatFromFileName outfile) outfile $ do
```
Yes, that is much better, done.

rvl commented on app/wallet-client/README.md:
```diff
@@ -0,0 +1,40 @@
+
+### How to use
+
+Start a wallet using `stack exec cardano-node -- ....`
```
Updated

rvl commented on src/Cardano/Wallet/Client/CLI/ProcessUtil.hs:
```diff
@@ -0,0 +1,56 @@
+interruptSelf :: IO ()
+interruptSelf = pure ()
+
+#endif
```
Running the CLI on Windows (as part of tests) is actually one of the reasons I wrote this code. The `wait-for-sync --pid` option won't work yet on Windows so I have marked it as TODO.

rvl commented on src/Cardano/Wallet/Client/HttpsSettings.hs:
```diff
@@ -0,0 +1,71 @@
+        }
+    clientSupported = def
+        { supportedCiphers = ciphersuite_default
+        }
```
Yes, it's the `AuthenticateServer` option. They should be de-duplicated. Will the `mkHttpsManagerSettings` function stay in `cardano-sl` or move elsewhere?

rvl commented on src/Cardano/Wallet/Client/Easy.hs:
```diff
@@ -0,0 +1,167 @@
+              (unSyncPercentage progress)
+              (fst info) (fromMaybe 0 (snd info))
+        info = ( unBlockchainHeight $ nfoLocalBlockchainHeight resp
+               , unBlockchainHeight <$> nfoBlockchainHeight resp )
```
Yes it does the same thing, plus logging the timing information, handling Ctrl-C, and giving up if a PID has gone. The wait function uses `Control.Retry`, but I'm not sure if that helped.

rvl commented on src/Cardano/Wallet/Client/CLI/Options.hs:
```diff
@@ -0,0 +1,251 @@
+clientAuthPemP = pemConfig <$> strOption (long "pem"
+                                          <> metavar "FILENAME"
+                                          <> help "Combined X509 certificate and key file")
+  where pemConfig pem = ClientAuthConfig pem pem
```
I want this program to inter-operate well with users' configurations. If they have generated certificates themselves then this option will help. In any case, it's easier to supply just `--pem` instead of both `--cert` and `--key`.

rvl commented on src/Cardano/Wallet/Client/Wait.hs:
```diff
@@ -0,0 +1,152 @@
+  SyncErrorTimedOut a    == SyncErrorTimedOut b    = a == b
+  SyncErrorInterrupted   == SyncErrorInterrupted   = True
+  SyncErrorException _   == SyncErrorException _   = True
+  _ == _ = False
```
I have now enabled EditorConfig in Emacs.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+run :: PlotOptions -> IO ()
+run PlotOptions{..} = do
+  df <- loadRecords poInputFile
+  plotRecords poPlot poExtraTitle poPlotLeft poPlotRight poOutputFile df
```
Done.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+data Record = Record Double [Double] deriving (Show)
+data DataFile = DataFile
+  { dfRecords   :: [Record]
+  , dfStartTime :: Text
```
Done.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+  parseJSON = withArray "Record" $ \v ->
+    if V.length v == 2
+      then (Record <$> parseJSON (v ! 0) <*> parseJSON (v ! 1))
+      else fail "Incorrect record array length"
```
Good idea, done.

rvl commented on src/Cardano/Wallet/Client/CLI/Options.hs:
```diff
@@ -0,0 +1,251 @@
+data Action m = WaitForSync (Maybe ProcessID) (Maybe FilePath)
+              | WaitForRestore (Maybe ProcessID) (Maybe FilePath)
+              | forall a. ToJSON a => WalletEndpoint (WalletCall m a) -- ^ Wallet API call
+              | WalletEndpointVoid (WalletCallVoid m) -- ^ Wallet API call with empty response
```
Done.

rvl commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -103,7 +103,7 @@ module Cardano.Wallet.API.V1.Types (
   , EstimatedCompletionTime
   , mkEstimatedCompletionTime
-  , SyncThroughput
+  , SyncThroughput(..)
```
What would be the best way to log the throughput value as a number that can be stored in a JSON file and then later plotted?

rvl commented on src/Cardano/Wallet/Client/Wait.hs:
```diff
@@ -0,0 +1,152 @@
+      status (Just e) = [ "success" .= False, "error" .= String (show e) ]
+
+instance ToJSON SyncError where
+  toJSON e = String (show e)
```
I don't think it's possible because there is no `Generic` instance of `SomeException`.

rvl commented on src/Cardano/Wallet/Client/CLI/Options.hs:
```diff
@@ -0,0 +1,251 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE LambdaCase                #-}
+
+module Cardano.Wallet.Client.CLI.Options where
```
It could, but I have a unit test for a function in this module.

Cabal new-build internal libraries might provide a way of keeping non-core wallet code out of the main library.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
```
:smile: Is `RecordWildcards` a problem in this file of this repo? The data type and usage are in the same file, and the record fields have prefixes.

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+  parseJSON = withArray "Record" $ \v ->
+    if V.length v == 2
+      then (Record <$> parseJSON (v ! 0) <*> parseJSON (v ! 1))
+      else fail "Incorrect record array length"
```
I intentionally wrote instances for this because they are short and the JSON format should look like a normal JSON data file, not something weird and haskelly. Is it OK? (the alternative is to define `genericParseJSON customOptions`)

rvl commented on app/sync-plot/Main.hs:
```diff
@@ -0,0 +1,116 @@
+    inputP = argument str (metavar "INFILE" <> help "Input JSON")
+    outputP = argument str (metavar "OUTFILE" <> help "Output PNG")
+
+data Record = Record Double [Double] deriving (Show)
```
Done.

rvl commented on src/Cardano/Wallet/Client/Easy.hs:
```diff
@@ -0,0 +1,167 @@
+
+data ConfigError = ConfigCredentialsError Text
+                 | ConfigCertificateError FilePath Text
+                 deriving (Show, Typeable)
```
Fixed.

rvl commented:
@ktorz @parsonsmatt Thanks for the reviews, I apologise for the delay (leave, other things, etc). I have rebased on develop and addressed/responded to your comments so far.

I also implemented a few more CLI commands for API endpoints (although the entire API is not covered yet).

About code formatting (everyone's favourite topic I know). I have fixed it, but are half-indented `where`s acceptable? That means the actual bindings are indented 4 spaces instead of 8.

What else should be done for this PR?

KtorZ commented:
@rvl Apologize for the time it takes but priority has been quite elsewhere at the moment for us. And this is extra code to maintain and test (even though, mind you, it is much appreciated). 
Since it's also a bit out-of-the-loop for us, we haven't really included this in our QA pipeline which should be done at some point. 

Might I can put that on @piotr-iohk radar. Would be a nice idea to play with it and see if things are working smoothly. This is of course of a lower priority vs anything about decoupling or BIP-44 :+1:

rvl commented:
OK I'll follow up with QA and see how I can assist them.
This tool is something that I need for devops work, because there is currently no easy way to use the wallet API.
Additionally, I have built the Windows chain sync acceptance tests on top of this program. We really need automated sync tests running on Windows.
What's the best way to get this merged? We don't want to have a long-running feature branch.

KtorZ commented:
I understand and I do agree with you that the situation is not practical here. What I am truly afraid of here is that, as soon as this lands on this repository, it becomes (eventually) the wallet's team responsibility. And that is something that comes a bit "out of nowhere" for us because it wasn't planned nor discussed with any of us :disappointed: 
We can't just afford this extra work at the moment. My recommendation here would be to have that maybe as _another repository_, maintained and managed by devops or, maybe a dedicated team later once we've agreed on spending time and resources on building and improving such tool.

KtorZ commented:
Closing as per last comment + private conversation.

KtorZ commented on src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -284,9 +292,22 @@ instance
+                \your mnemonic sentence."
+            ErrEntropy ErrInvalidEntropyLength{} ->
+                "Something went wrong when trying to generate the entropy from \
+                \the given mnemonic. As a user, there's nothing you can do."
```
Note that I am almost tempted to shove in a `error "OMG SOMETHING WENT WRONG"` here because really, we can never reach this branch. It'd mean we'd bypassed our `Mnemonic` type, which, as far as I am concerned, isn't possible currently.

piotr-iohk commented on src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -284,9 +292,22 @@ instance
+            ErrMnemonicWords ErrWrongNumberOfWords{} ->
+                "Invalid number of words: at least "
+                <> show (natVal (Proxy :: Proxy mw))
+                <> " words are expected."
```
Would it always say `at least 15` for `mnemonic_sentence` and `at least 9` for `mnemonic_second_factor` ?

KtorZ replied:
Indeed.

piotr-iohk replied:
Also for `mnemonic_second_factor` (where we allow 9 or 12) it will show somewhat faulty message for, say, 15 words I suppose...

piotr-iohk replied:
Maybe it could be a bit more comprehensive, like:
`Wrong number of words - N. Supported number of words are: mnemonic_sentence [15,18,21,24]; mnemonic_second_factor [9,12]`

KtorZ replied:
Hmm.. No :) 
The mnemonic length is actually encoded at the type-level here (that's what `mw` symbolizes) and, the error message is constructed based on this type. So, for _any_ representation `[9,12]` or `[15,18,21,24]` or `[12,18]` and so forth, the error will always report on the first (smallest) length.

piotr-iohk replied:
>Also for mnemonic_second_factor (where we allow 9 or 12) it will show somewhat faulty message for, say, 15 words I suppose...

I mean that when user provides 15 words for `mnemonic_second_factor` he will get `Invalid number of words: at least 9 words are expected`... well 15 is more than 9 so that's _somewhat misleading_

That's why I suggested:
`Wrong number of words - N. Supported number of words are: mnemonic_sentence [15,18,21,24]; mnemonic_second_factor [9,12]`

That is if you can throw in N (if not, we can drop it)

KtorZ replied:
Ah, got you! That requires slightly more work than it seems actually. Because of the way operations at the type-level work on list (at least, here), we do define instance in an inductive / recursive manner and, we "loose" (we actually consume them) some type-level information as we do it. I can try something for a few minutes, but let's scope it to 15 minutes ^^"

piotr-iohk reviewed (approved):
Nice one :1st_place_medal:  :)
